File: typescript/src/Line.ts
import { Point } from './Point';

export class Line {
    constructor(public readonly a: Point, public readonly b: Point) {}

    get rise(): number {
        return this.b.y - this.a.y;
    }

    get run(): number {
        return this.b.x - this.a.x;
    }

    get length(): number {
        return Math.sqrt(this.rise ** 2 + this.run ** 2);
    }
}

File: rust/tests/geometry/line.rs
use crate::geometry::{line::Line, point::Point};

#[test]
fn test_line_calculate_rise() {
    let start = Point { x: 1.0, y: 2.0 };
    let end = Point { x: 3.0, y: 5.0 };
    let line = Line { start, end };
    
    assert_eq!(line.rise(), 3.0);
}

#[test]
fn test_line_calculate_run() {
    let start = Point { x: 1.0, y: 2.0 };
    let end = Point { x: 3.0, y: 5.0 };
    let line = Line { start, end };
    
    assert_eq!(line.run(), 2.0);
}
